Migrate project to BothamNetworking.

pedrovgs opened the pull request:
Fix #9 fix #8 fix #5.

This PR is not going to compile or pass the tests in Travis environments because BothamNetworking has not been published yet. I've changed the project repository to a private on to be able to merge this branch into master and the rest of the pending branches before the first release even if the project is not passing in travis.

We should create a new target inside the BothamNetworking project with the SwiftyJson dependency and the json parsing implementation based on Result extension. What do you think?

pedrovgs commented on MarvelApiClient/Result.swift:
```diff
@@ -0,0 +1,31 @@
+import Result
+import SwiftyJSON
+
+extension ResultType where Value == HTTPResponse {
```
Ahhhhhh I was trying with Result instead of ResultType!!! Thanks Davide! Your Swift-Fu is over 9000!

![Over9000](http://orig08.deviantart.net/8580/f/2013/197/6/b/vegeta_shocked_9000_by_wortmann-d6drw47.gif)

pedrovgs commented on MarvelApiClient/CharactersAPIClient.swift:
```diff
@@ -0,0 +1,47 @@
+    public func getAll(offset offset: Int, limit: Int,
+        completion: (Result<GetCharactersDTO, BothamAPIClientError>) -> ()) {
+        assert(offset >= 0 && limit >= 0)
+        let params: [String:String?] =  [MarvelAPIParams.offset : "\(offset)", MarvelAPIParams.limit : "\(limit)"]
```
I remember why we changed this. This is the ``Alamofire`` implementation:

```swift
// MARK: - Request

    /**
        Creates a request for the specified method, URL string, parameters, parameter encoding and headers.
        - parameter method:     The HTTP method.
        - parameter URLString:  The URL string.
        - parameter parameters: The parameters. `nil` by default.
        - parameter encoding:   The parameter encoding. `.URL` by default.
        - parameter headers:    The HTTP headers. `nil` by default.
        - returns: The created request.
    */
    public func request(
        method: Method,
        _ URLString: URLStringConvertible,
        parameters: [String: AnyObject]? = nil,
        encoding: ParameterEncoding = .URL,
        headers: [String: String]? = nil)
        -> Request
    {
        let mutableURLRequest = URLRequest(method, URLString, headers: headers)
        let encodedURLRequest = encoding.encode(mutableURLRequest, parameters: parameters).0
        return request(encodedURLRequest)
    }
```

Right now I'd like to change the API to use [String:String] instead of [String:String?] but this is just my opinion and the code is working. This is something we can discuss with the rest of the team.

davideme replied:
I'm fine if we change in the library if we consider as default behaviour, and we open a ticket to implement as a config, that:
// By Default
["var1" : ""] generate var1=
// and set 
["var1" : ""] generate var1

pedrovgs replied:
Can you open an issue describing this?

pedrovgs commented on MarvelAPIClientPlayground.playground/contents.xcplayground:
```diff
@@ -1,4 +1,4 @@
 <?xml version="1.0" encoding="UTF-8" standalone="yes"?>
-<playground version='5.0' target-platform='ios' display-mode='rendered'>
```
Sorry, I forgot to revert this change. Fixed in my last commit.

pedrovgs commented on MarvelApiClient.xcworkspace/xcshareddata/MarvelApiClient.xcscmblueprint:
```diff
@@ -0,0 +1,30 @@
+{
```
Fixed!

pedrovgs commented on MarvelApiClient/DefaultHeadersRequestInterceptor.swift:
```diff
@@ -0,0 +1,19 @@
+import Foundation
+import BothamNetworking
+
+class DefaultHeadersRequestInterceptor: BothamRequestInterceptor {
```
My interceptor implementation or the interceptor mechanism?

davideme replied:
Interceptor mechanism, that to set Headers you have yo implement an interceptor,

Maybe if we see that a recuring feature we can define DefaultHeadersRequestInterceptor as part of Botham, and have as param or method of HTTPClient.

pedrovgs replied:
IMHO the interceptor mechanism is quite cool :) As you say, we can provide this implementation by default as part of BothamNetworking.

pedrovgs commented on MarvelApiClient/MarvelAPIAuthentication.swift:
```diff
@@ -0,0 +1,31 @@
+            MarvelAPIParams.hash : hash]
+        return request.appendingParameters(authParams)
+    }
+
```
I don't think so, we don't need it right now. The ``BothamAPIClient`` is already returning an ``BothamAPIClientError.HTTPResponseError`` with the status code if the authentication is not working.

davideme replied:
Yeah, we but have to review that, we need a way to capture authentication error.

pedrovgs replied:
With the current BothamNetworking code we can implement all the "retry request to authenticate" mechanism from the BothamAPIClient users. If we want to implement this mechanism inside interceptors we will have to improve interceptors implementation to follow a chain of responsibility pattern. Fortunately we don't need it for this marvel api client :)

davideme replied:
let's skip this for now. we will review it when implementing authentication.

pedrovgs commented on MarvelApiClient/MarvelAPIClientConfig.swift:
```diff
@@ -8,9 +8,9 @@
 
     static let host = "http://gateway.marvel.com/v1/public/"
-    static let defaultHeaders = ["Accept" : "application/json"]
+    static let defaultHeaders = ["Accept" : "application/json", "Content-Type" : "application/json"]
```
Nop.

davideme replied:
Ok, we have something inconsistent.
https://github.com/Karumi/BothamNetworking/blob/34feaa97c2815c60d4b4d5035c9a9b860558f9a0/BothamNetworking/HTTPClient.swift#L32

We force check that the return content is JSON, but we don't send "Accept" : "application/json".

And for the Request Encoding+Content-Type maybe we should be able to configure it or provide one by default.

pedrovgs replied:
I've removed that check from the BothamNetworking project. This is the PR https://github.com/Karumi/BothamNetworking/pull/23

davideme replied:
As discussed outside of this PR we will remove the check from the BothamNetworking.

pedrovgs commented on MarvelApiClient/Result.swift:
```diff
@@ -0,0 +1,31 @@
+
+    private func dataToJSONResult(data: NSData?) -> Result<JSON, Error> {
+        return Result<JSON, Error> {
+            let object: AnyObject = try NSJSONSerialization.JSONObjectWithData(data ?? NSData(),
```
Done! Review my last commit, my implementation is a little bit different, but it is also working :)

pedrovgs commented:
@davideme as you suggested I've updated ``DefaultHeadersRequestInterceptor`` to do change any header value previously configured in the ``HTTPRequest`` creation. You can take a look at the new implementation in the test I've added heer ``DefaultHeadersRequestInterceptorTests``

pedrovgs commented:
@davideme I've updated this PR to match with the latests BothamNetworking changes. If I'm not wrong, this is ready to merge.

davideme commented:
:+1:
